### ISSUE TITLE
bitcoin core json-rpc interface with pruning enabled

### ISSUE THREAD

**Issue by chris-belcher** [JoinMarket-Org/joinmarket#177]
We should be mindful about the blockchain pruning feature of Bitcoin Core, and build our systems to be able to use it. Ideally we'd like to be able to use JoinMarket with pruning enabled. Note that right now in Bitcoin 0.11 the wallet is disabled if pruning is enabled so we can't use it anyway.

Right now we use sync_addresses() which finds all the addresses that have been used already, followed by sync_unspent() which gets all the UTXOs. Historically it works like this because the blockr.io interface needed to learn which addresses were used so it could put them in a /unspent/ http request to find UTXOs. But there's no reason the Bitcoin Core interface has to work this way.

We could simply get the most recently used address by finding the ones that has UTXOs on them (or using the index_cache from the wallet file if that field exists)

Pruning implies -rescan won't work. I understand the bitcoin core developers don't want software to rely on -rescan too much anyway.

So if we found a way to go from address to UTXOs then we could do this. One way would be to simply use RPC calls getbestblockhash, getblock and gettxout to step through all outputs in the blockchain, checking if our addresses match anything. This might take ages but we could cache it into some database.
According to phantomcircuit on IRC, this method is a bit silly and it would be better to just write the RPC call that dumps UTXO-set data in the manner of listtransactions and open a pull request to bitcoin core.
Either way, pruning is still not really finished as a feature in bitcoin core so it's worth waiting to see how it turns out, or contributing ourselves.

**Comment by chris-belcher** [comment]
BTW pruning also breaks Armory so it's worth looking at how they handle it.

**Comment by chris-belcher** [comment]
Need to read some of the old bitcoin-dev mailing list posts about pruning.

**Comment by neocogent** [comment]
I'd like to bring your attention to my sqlChain project. This is a mysql layer over top bitcoind. It can run on a pruning node and manages the node so that it syncs the data to mysql before it gets pruned, minimizing storage demands. There are a number of options and two api layers I have already working on top of it. One provides Insight/BCI/Websocket compatible json apis and the other an Electrum server (currently private use, without peer discovery)

I'm a bit ahead of myself now as I want to do more testing before it gets out there and known. I just wanted you to be aware of it and maybe check it out to see how it could work with joinmarket in future. It doesn't currently have a wallet api but it could be a next step, or by using Elecrum with sqlChain as a private server, you get the advantage of HD wallet with your own piece of mind for trust (fast enough to not need utxo set download).

On top a pruning node, with Electrum server, it currently needs about 28GB total disk space (including bitcoind), and with that you get full Insight/BCI API + websocket, to boot (optionally).

My blog is [www.neocogent.com](http://www.neocogent.com), where I discuss programing sqlChain, and my [github](https://github.com/neocogent/sqlchain) has actively committed full open source. I've been working on this for the last 2-3 weeks, so it's very new, and plan to start up an AWS demo api server to show off how well it works, while testing in that environment. sqlChain has both an utxo api and a getMerkle api (needed by Electrum clients). 

Anyway - just a heads up, as it's related to this topic.

**Comment by chris-belcher** [comment]
@neocogent for a database of key-value mappings between addresses and UTXOs, how large would it be?

**Comment by neocogent** [comment]
My current db stats, now at 370,138 (last night I stopped syncing) are:

- addresses 89,734,757 = 3.7GB
- outputs 240,418,498 = 16.3 GB
- trxs 80,118,383 = 6.3 GB (+ 2 up to 27 GB blob data depending on --no-sigs option)
- blocks 370,138 = 33MB (+3MB hdr data)

This is the full blockchain (including indexes). So if you  were only tracking specific wallet addresses, which currently sqlchain doesn't have an option for, then that would be tiny in comparison. You couldn't store outputs (utxo set) without having the other tables as they are all inter-dependent (related). With this data the full insight and BCI apis, and Electrum server api (except getProof one, which appears to not be used anyway), can be served.

I'm still figuring out how well this will actually work on a pruning node. It's a bit unclear to me what validation a pruning node does before discarding blocks. I guess it cannot fully validate each tx but can check the tx hash is valid against header/merkle data. I don't think it maintains it's own utxo set. sqlchain does keep that data but right now it doesn't verify txs itself (only decodes what bitcoind gives it and checks headers).

**Comment by neocogent** [comment]
Oh, I realized something after I wrote above. The outputs table contains all outputs, even spent ones. Right now I don't prune that but it would be feasible in a different use of the db. I think, based on some queries I ran on the db, that only about 10% of the outs are unspent so the outputs table could probably be reduced to ~1.6GB if spent outs were pruned. In that case you could manage a wallet fully on sqlchain with only 1.6GB outputs and ~10GB for addresses and trxs. Or less, as a siginificant chunk of old addresses unrelated to utxos could also be pruned. So, in theory a fully validated wallet version of sqlchain could run with maybe a fair bit less than 10GB.

I was trying to figure out just what a pruning does to validate txs. From what I read by nullc it does truly validate txs as good as a full node, but I know it prunes old blocks. So I guess it does keep the utxo set. I'm just not sure where it stores that or how much space it uses. I did a check on my test pruning node and it's only using about 440 MB more than the blocks kept in chainstate - so I guess that's what it uses to validate. I was a bit unclear but I thought it did fully validate txs.

**Comment by neocogent** [comment]
After syncing again I did a bit of querying the db.

There are ~33M unspent outputs out of 241M total, using about 5M addresses out of 90M total. This leads me to believe the unspent output set would take about 2.2 GB and the address index to match about 205MB. So say about 2.5 GB right now. I'd guess not the most compact method for this but it is fast, and this has tx value (in satoshis) and tx id for that table.

**Comment by chris-belcher** [comment]
@neocogent Unfortuntely I don't think your database is appropriate. It would be yet another dependency and it would be quite overkill, the database contains far more information than we need.

**Comment by chris-belcher** [comment]
Once you've got your joinmarket wallet sync'd (how exactly we'll get to) then everything runs fine. Calls to gettxout, gettransaction, sendrawtransaction, getaccount, importaddress etc all work.

Actually sync'ing is the problem. Two possible ways to do it.

1. Sync'ing wallet the first time requires importing many watch-only addresses and redownloading the entire blockchain, only to delete most it by pruning. After that everything would work the same way as now. I don't much like the idea of redownloading the blockchain, potentially many times as we don't know how many addresses to import.

1. Sync'ing the wallet only queries the UTXO set. The transaction history is not known. JoinMarket would be able to tell you the balance of a wallet and it's UTXO but not why the balance is that way. Once money is spent the wallet would forget the payment ever existed. The position in the HD key chain could be found by searching a number of addresses, displaying the total btc amount found and asking the user "Look ahead more?". Presumably the user knows how much money they had and/or can decide on whether they have the patience to keep looking ahead. (The amount of looking ahead can increase with each cycle of looking ahead.)
The next address to use for receiving coins could just be the index after the last address with UTXOs on it. Although there is a danger that it ends up reusing addresses if this address has spent TXs on it.

An issue now with this is that the wallet is disabled for pruning nodes so -walletnotify wont work.

**Comment by chris-belcher** [comment]
https://bitcointalk.org/index.php?topic=1169146.msg12314549

**Comment by neocogent** [comment]
I can't comment on bitcointalk so thought I'd just add something here. I was able to use joinmarket with bitcoin core without doing a full rescan. If you know the addresses are new (no history) then this works ok. The problem is I don't recall how I did it. Most probably I did the import and it started a rescan and then I just stopped bitcoind, and then started it again without rescan. Unfortunately I'm not sure now as I didn't keep notes but I do recall after that I ran joinmarket and it behaved properly checking addresses and I didn't have to wait ages for the rescan. So it may be worth checking how if this reset/restart process works. In most cases you will always know that new addresses in joinmarket will have no history, and the rescan is pointless.
